treeherder/auth/backends.py

```diff
@@ -1,93 +1,170 @@
+
+        if not auth:
+            raise AuthenticationFailed("Authorization header is expected")
+
+        parts = auth.split()
+
+        if parts[0].lower() != "bearer":
+            raise AuthenticationFailed("Authorization header must start with 'Bearer'")
+
+        elif len(parts) == 1:
+            raise AuthenticationFailed("Token not found")
+
+        elif len(parts) > 2:
+            raise AuthenticationFailed("Authorization header must be 'Bearer {token}'")
+
+        token = parts[1]
+
+        return token
+
+    def _get_username_from_userinfo(self, user_info):
         """
-        Extract the user's email from the client_id
+        Get the user's username from the jwt sub property
         """
 
-        # Client IDs must be in one of these forms:
-        # - email/foo@bar.com
-        # - mozilla-ldap/foo@bar.com
-        # Email regex taken from http://emailregex.com
-        match = CLIENT_ID_RE.match(client_id)
-        if match:
-            return match.group(1)
+        subject = user_info['sub']
+        email = user_info['email']
 
-        raise NoEmailException(
-            "No email found in clientId: '{}'".format(client_id))
+        if "Mozilla-LDAP" in subject:
+            return "mozilla-ldap/" + email
+        elif "email" in subject:
+            return "email/" + email
+        elif "github" in subject:
+            return "github/" + email
+        else:
+            raise AuthenticationFailed("Unrecognized identity")
 
-    def authenticate(self, request):
+    def _get_jwks_json(self):
+        """
+        Get the JSON Web Key Set (jwks), which is a set of keys
+        containing the public keys that should be used to verify
+        any JWT issued by the authorization server. Auth0 exposes
+        a JWKS endpoint for each tenant, which is found at
+        'https://' + AUTH0_DOMAIN + '/.well-known/jwks.json'. This endpoint
+        will contain the JWK used to sign all Auth0 issued JWTs for this tenant.
+        Reference: https://auth0.com/docs/jwks
+
+        The jwks is under our (Mozilla's) control. Changing it would be a big thing
+        with lots of notice in advance. In order to mitigate the additional HTTP request
+        as well as the possiblity of receiving a 503 status code, we use a static json file to
+        read its content.
+        """
+        cache_key = 'well-known-jwks'
+        cached_jwks = cache.get(cache_key)
 
-        auth_header = request.META.get("HTTP_TCAUTH", None)
-        host = request.get_host().split(":")[0]
-        port = int(request.get_port())
+        if cached_jwks is not None:
+            return cached_jwks
 
-        if not auth_header:
-            # Doesn't have the right params for this backend.  So just
-            # skip and let another backend have a try at it.
-            return None
+        cached_jwks = json.load(open('treeherder/auth/jwks.json'))
+        cache.set(cache_key, cached_jwks)
```

Comment:
If we're using a local file I think we should skip caching in Redis and load once at initial runtime?

Reply:
Caching or not, I think both are fine. I'll send a commit to remove caching.

Reply:
I think there are a few possible approaches (assuming embedding `jwks.json` in the codebase, which I think is a sensible tradeoff):

1. (the approach used here previously) for each incoming auth request, check if the jwks is cached, if so use that, otherwise load from disk, and then store to cache.
1. (the approach used after the PR update) for each incoming auth request, load jwks from disk
1. at process start, load the jwks into a constant at the top of this file, which is referenced by each request

These all have different trade-offs:
* Approach 1 makes a roundtrip to the remote Redis instance (it's not a localhost server or Django's locmem cache) for every request, to fetch a constant. Every N minutes (where N defaults to 5) it would have a cache miss and have to hit the disk again. 
* Approach 2 hits the disk for each request (the Heroku dynos most likely aren't great for disk I/O)
* Approach 3 means keeping the key permanently in memory (but given it's tiny I don't think it's even worth worrying about)

Approach 3 was what I was meaning above (I probably could have been clearer - sorry!) - but open to ideas? :-)

Reply:
Thanks for elaborating :) Approach 3 sounds great to me! I pushed a commit for that.